integration/test_collection.py

```diff
@@ -614,6 +705,104 @@ def test_mono_references_grcp(client: weaviate.Client):
     assert objects[0].data["ref"][0].data["ref"][1].data["name"] == "A2"
 
 
+def test_mono_references_grcp_typed_dicts(client: weaviate.Client):
+    client.collection.delete("ATypedDicts")
+    client.collection.delete("BTypedDicts")
+    client.collection.delete("CTypedDicts")
+
+    class AProps(TypedDict):
+        name: str
+
+    A = client.collection.create(
+        CollectionConfig(
+            name="ATypedDicts",
+            vectorizer=Vectorizer.NONE,
+            properties=[
+                Property(name="Name", data_type=DataType.TEXT),
+            ],
+        )
+    )
+    uuid_A1 = A.data.insert(AProps(name="A1"))
+    uuid_A2 = A.data.insert(AProps(name="A2"))
+
+    class BPropsInsert(TypedDict):
+        name: str
+        ref: ReferenceTo
+
+    B = client.collection.create(
+        CollectionConfig(
+            name="BTypedDicts",
+            properties=[
+                Property(name="Name", data_type=DataType.TEXT),
+                ReferenceProperty(name="ref", target_collection="ATypedDicts"),
+            ],
+            vectorizer=Vectorizer.NONE,
+        ),
+        BPropsInsert,
+    )
+    uuid_B = B.data.insert(BPropsInsert(name="B", ref=ReferenceTo(uuids=uuid_A1)))
+    B.data.reference_add(from_uuid=uuid_B, from_property="ref", ref=ReferenceTo(uuids=uuid_A2))
+
+    class CPropsInsert(TypedDict):
+        name: str
+        ref: ReferenceTo
+        age: int
+
+    C = client.collection.create(
+        CollectionConfig(
+            name="CTypedDicts",
+            properties=[
+                Property(name="Name", data_type=DataType.TEXT),
+                Property(name="Age", data_type=DataType.INT),
+                ReferenceProperty(name="ref", target_collection="BTypedDicts"),
+            ],
+            vectorizer=Vectorizer.NONE,
+        ),
+        CPropsInsert,
+    )
+    C.data.insert(CPropsInsert(name="find me", age=10, ref=ReferenceTo(uuids=uuid_B)))
+
+    class BPropsGet(TypedDict):
+        name: str
+        ref: Reference[AProps]
+
+    class CPropsGet(TypedDict):
+        name: str
+        ref: Reference[BPropsGet]
+        not_specified: str
+
+    objects = C.query.bm25_flat(
+        query="find",
+        return_properties=[
+            "name",
+            "age",
+            LinkTo(
+                link_on="ref",
+                properties=[
+                    "name",
+                    LinkTo(
+                        link_on="ref",
+                        properties=["name"],
+                        metadata=MetadataQuery(uuid=True),
+                    ),
+                ],
+                metadata=MetadataQuery(uuid=True, last_update_time_unix=True),
+            ),
+        ],
+        # type_=CPropsGet,
```

Comment:
Ah yes, this is because the gRPC refactoring isn't in this PR so I've left the generics in queries for later. We should also discuss this aspect as I don't remember a clear decision on the desired implementation